precompiles/utils/src/revert.rs

```diff
@@ -0,0 +1,352 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities to work with revert messages with support for backtraces and
+//! consistent formatting.
+
+use crate::{Bytes, EvmDataWriter};
+use alloc::string::{String, ToString};
+use fp_evm::{ExitRevert, PrecompileFailure};
+use sp_std::vec::Vec;
+
+/// Represent the result of a computation that can revert.
+pub type MayRevert<T = ()> = Result<T, Revert>;
+
+/// Generic error to build abi-encoded revert output.
+/// See: https://docs.soliditylang.org/en/latest/control-structures.html?highlight=revert#revert
+#[precompile_utils_macro::generate_function_selector]
+#[derive(Debug, PartialEq)]
+pub enum RevertSelector {
+	Generic = "Error(string)",
+}
+
+#[derive(Clone, PartialEq, Eq)]
+enum BacktracePart {
+	Field(String),
+	Tuple(usize),
+	Array(usize),
+}
+
+/// Backtrace of an revert.
+/// Built depth-first.
+/// Implement `Display` to render the backtrace as a string.
+#[derive(Default, PartialEq, Eq)]
+pub struct Backtrace(Vec<BacktracePart>);
+
+impl Backtrace {
+	/// Create a new empty backtrace.
+	pub fn new() -> Self {
+		Self(Vec::new())
+	}
+
+	/// Check if the backtrace is empty.
+	pub fn is_empty(&self) -> bool {
+		self.0.is_empty()
+	}
+}
+
+impl core::fmt::Display for Backtrace {
+	fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+		for (i, part) in self.0.iter().rev().enumerate() {
+			match (i, part) {
+				(0, BacktracePart::Field(field)) => write!(f, "{field}")?,
+				(_, BacktracePart::Field(field)) => write!(f, ".{field}")?,
+				(_, BacktracePart::Tuple(index)) => write!(f, ".{index}")?,
+				(_, BacktracePart::Array(index)) => write!(f, "[{index}]")?,
+			}
+		}
+		Ok(())
+	}
+}
+
+/// Possible revert reasons.
+#[non_exhaustive]
+#[derive(PartialEq, Eq)]
+pub enum RevertReason {
+	/// A custom revert reason if other variants are not appropriate.
+	Custom(String),
+	/// Tried to read data out of bounds.
+	ReadOutOfBounds {
+		/// What was being read?
+		what: String,
+	},
+	/// An unknown selector has been provided.
+	UnknownSelector,
+	/// A value is too large to fit in the wanted type.
+	/// For security reasons integers are always parsed as `uint256` then
+	/// casted to the wanted type. If the value overflows this type then this
+	/// revert is used.
+	ValueIsTooLarge {
+		/// What was being read?
+		what: String,
+	},
+	/// A pointer (used for structs and arrays) points out of bounds.
+	PointerToOutofBound,
+	/// The reading cursor overflowed.
+	/// This should realistically never happen as it would require an input
+	/// of length larger than 2^64, which would cost too much to be included
+	/// in a block.
+	CursorOverflow,
+	/// Used by a check that the input contains at least N static arguments.
+	/// Often use to return early if the input is too short.
+	ExpectedAtLeastNArguments(usize),
```

Comment:
Could call this `ExpectedMoreArguments`

Reply:
What would be the message instead of `"Expected at least {n} arguments"` ?